crates/worker/src/partition_processor_manager.rs

```diff
@@ -684,68 +763,80 @@ impl<T: TransportConnect> PartitionProcessorManager<T> {
+                    RunMode::Leader
+                } else if control_processor.command == ProcessorCommand::Follower {
+                    RunMode::Follower
                 } else {
-                    debug!("Unknown partition id '{partition_id}'. Ignoring follower command.");
-                }
-            }
-            ProcessorCommand::Leader => {
-                if let Some(state) = self.running_partition_processors.get_mut(&partition_id) {
-                    state
-                        .run_for_leader(
-                            self.metadata_store_client.clone(),
-                            self.metadata.my_node_id(),
-                        )
-                        .await?;
+                    unreachable!()
+                };
+
+                if let Some(status) = self.running_partition_processors.get_mut(&partition_id) {
+                    if let ProcessorStatus::Started(state) = status {
+                        // if we error here, then the system is shutting down
+                        if run_mode == RunMode::Follower {
+                            state.step_down()?;
+                        } else if run_mode == RunMode::Leader {
+                            state
+                                .run_for_leader(
+                                    self.metadata_store_client.clone(),
+                                    self.metadata.my_node_id(),
+                                )
+                                .await?;
+                        }
+                    } else {
+                        // otherwise if it's still starting we can't control it yet
+                        debug!(%partition_id, "Ignoring control partition processor message. Partition processor is still starting");
+                    }
                 } else if let Some(partition_key_range) = partition_table
                     .get_partition(&partition_id)
                     .map(|partition| &partition.key_range)
                 {
-                    self.start_partition_processor(
+                    let starting_task = self.start_partition_processor_task(
                         partition_id,
-                        partition_key_range,
-                        RunMode::Leader,
-                    )
-                    .await?;
+                        partition_key_range.clone(),
+                        run_mode,
+                        events,
+                    );
+
+                    // note: We starting each task on it's own thread due to an issue that shows up on MacOS
+                    // where spawning many tasks of this kind causes a lock contention in tokio which leads to
+                    // starvation of the event loop.
+                    let handle = self.task_center.spawn_blocking_unmanaged(
+                        "starting-partition-processor",
+                        Some(partition_id),
+                        starting_task.run().instrument(
```

Comment:
Maybe we can limit the concurrency of the number of starting tasks using a `Semaphore` to work around the problem I encountered on my machine?